### DISCUSSION TITLE
Mobile onSuccess()

### DISCUSSION THREAD

**Discussion by atix07** [wevm/wagmi#2067] [Q&A]
Hello! I have so many trouble on mobile phones... One of them is onSuccess promise when transaction is confirmed/succeed. 

Example:
```
const {
    isLoading: loadingBuy, isSuccess, write: buyWithUsdt,
  } = useContractWrite({
    ...presaleContract,
    functionName: 'buyWithUSDT',
    args: [coinAmount],
    overrides: {
      from: userAddress,
    },
    onSuccess(res) {
      sethash(res.hash);
      setAllowBack(false);
      setAllowWait(false);
    },
  });
```
I tried making write async with using .then() same happens.

On success I want to show "Thank you" message but I think its timeouting when switching between wallet and website. Its working 50% of the time. Really not great user experience. What can I do for it to improve? Thank you

**Comment by tmm** [comment]
You should use [`useWaitForTransaction`](https://wagmi.sh/react/hooks/useWaitForTransaction) `onSuccess` instead.

**Reply by atix07** [reply]
> You should use [`useWaitForTransaction`](https://wagmi.sh/react/hooks/useWaitForTransaction) `onSuccess` instead.

Thank you I will play around with useWaitForTransaction. I assume data.hash becomes available even thought it takes some time to redirect back?

**Reply by atix07** [reply]
> You should use [`useWaitForTransaction`](https://wagmi.sh/react/hooks/useWaitForTransaction) `onSuccess` instead.

Sadly it doesnt make diffrence. Like onSuccess(), data.hash is not available after redirected back to website from wallet. 

```
const {
    data: buyData, isLoading: loadingBuy, isSuccess, write: buyWithUsdt,
  } = useContractWrite({
    ...presaleContract,
    functionName: 'buyWithUSDT',
    args: [coinAmount],
    overrides: {
      from: userAddress,
    },
  });

  const { isLoading: loadingTransBuy } = useWaitForTransaction({
    hash: buyData?.hash,
    onSuccess() {
      sethash(buyData?.hash);
      setAllowBack(false);
      setAllowWait(false);
    },
    onError() {
      setFeedback('Transaction failed');
    },
  });
```

I would say its mobile limitation. If it passes 300ms on redirect, nothing happens.
